source/AI.cpp

```diff
@@ -3514,13 +3514,56 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 	AimTurrets(ship, firingCommands, !Preferences::Has("Turrets focus fire"));
 	if(Preferences::Has("Automatic firing") && !ship.IsBoarding()
 			&& !(autoPilot | activeCommands).Has(Command::LAND | Command::JUMP | Command::FLEET_JUMP | Command::BOARD)
-			&& (!target || target->GetGovernment()->IsEnemy()))
+			&& (!target || target->GetGovernment()->IsEnemy())
+			&& !isMouseTurningEnabled)
 		AutoFire(ship, firingCommands, false);
+
+	if (activeCommands.Has(Command::MOUSETURNING)) {
+		isMouseTurningEnabled = !isMouseTurningEnabled;
+	}
+
+	if (isMouseTurningEnabled && !ship.IsBoarding() && !ship.IsReversing())
+	{
+		Angle shipAngle = ship.Facing();
+		if (rightMouseButtonHeld)
+		{
+			int index = 0;
+			for(const Hardpoint &hardpoint : ship.Weapons())
+			{
+				if(hardpoint.IsReady() && !hardpoint.GetOutfit()->Icon())
+					firingCommands.SetFire(index);
+				++index;
+			}
+		}
+
+		// To stop jittering due to the mouseAngle changing
+		// frequently, the ship is only turned if both the
+		// angles differ by more than or equal to one degree
+		if (abs(mouseAngle.Degrees360() - shipAngle.Degrees360()) < 1 )
+		{
+			command.SetTurn(0);
+		}
+		else if (abs(mouseAngle.Degrees360() - shipAngle.Degrees360()) < 180 )
+		{
+			if (shipAngle.Degrees360() < mouseAngle.Degrees360())
+				command.SetTurn(1);
+			if (shipAngle.Degrees360() > mouseAngle.Degrees360())
+				command.SetTurn(-1);
+		}
+		else
+		{
+			if (shipAngle.Degrees360() > mouseAngle.Degrees360())
+				command.SetTurn(1);
+			if (shipAngle.Degrees360() < mouseAngle.Degrees360())
+				command.SetTurn(-1);
+		}
```

Comment:
I feel like this logic can be compacted but I'm not really sure how exactly ; just giving a direction that matches the difference between the two angles and not checking every single case
I'm not sure how much the fact that these degrees are between 0 and 360 instead of -180 and 180 is helping for that either.